docs/signature-layout.md

```diff
@@ -0,0 +1,35 @@
+# Signature File Layout
+
+A common file layout for storing and serving signatures provides a consistent way to reference image signatures. Signatures on a filesystem or a web server shall use this common layout. Signatures stored in a REST API are not required to use this common layout.
+
+## Specification
+
+This specification relies on [RFC3986](https://tools.ietf.org/html/rfc3986), focusing on defining a [path component](https://tools.ietf.org/html/rfc3986#section-3.3) to compose a concise URI reference to a signature.
+
+**SCHEME[AUTHORITY]/PATH_PREFIX/IMAGE@MANIFEST_DIGEST/signature-INT**
+
+**Definitions**
+
+* **SCHEME**: URI scheme per [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.1), e.g. **file://** or **https://**
+* **AUTHORITY**: An optional authority reference per [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.2), e.g. **example.com**
+* **PATH_PREFIX**: An arbitrary base path to the image component
+* **IMAGE**: The name of the image per [v2 API](https://docs.docker.com/registry/spec/api/#/overview). This would typically take the form of registry/repository/image but is not required to have exactly three parts. There is no requirement to include a **:PORT** component but it should be included if part of the image reference.
```

Comment:
On Wed, Oct 26, 2016 at 02:12:21PM -0700, Miloslav Trmač wrote:
> And there is a specific bad failure case: removing signatures..

With signatures in the Merkle tree, the addition or removal of signatures changes the hash of the Merkle root.  So it doesn't matter what A does, B still asserts that sha256:123… (containing both a B sig and the RH sig) is `rhel/rhel7`.  A client attempting to fetch the image will check as many ref engines as it wants (e.g. the ref-engine for A's repository and the ref-engine for B's repository.  It sees the A ref-engine claiming `rhel/rhel7` is sha256:abc… and the B ref-engine claiming `rhel/rhel7` is sha256:123….  It walks both trees to find signatures, only finds acceptable signatures on sha256:123…, and proceeds to unpack it.

With a single ref-engine shared by all tenants, the registry could maintain a signature-list blob at the root of the Merkle tree with all the signatures that anyone ever wanted to apply to the name-assertion (e.g. `{"name": "rhel/rhel7", "blob": {"digest": "sha256:123…", …}}`).  Then the name-assertion is a stable blob that can be signed, and the registry keeps the signature list, only granting revocations to folks who can prove ownership of the signing key.  In this case, A wouldn't be allowed to remove the RH signature.  The RH folks could remove the RH signature, and if they wanted to do that then B would *appropriately* consider the sha256:123… image unacceptable.  If that is too opaque, the registry could keep the signature list append-only and also keep a revocation list (which B could respect or ignore as it saw fit).

> In a first version, the system would probably prohibit removing signatures at all, and every customer would see 50 (or 10000?) signatures on the shared `rhel/rhel7` image which they don’t care about and which would never be removed

If you get too many signatures by using a very popular ref engine, setup your own curated ref engine with your own signature-curation rules for the root signature-list.  This is the same as the A and B ref engine case above, and you can still keep all the blobs from the name-assertion on down in the central CAS engine.